File: android/src/main/java/in/cashify/otaupdate/OtaModuleManager.kt
package `in`.cashify.otaupdate

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

/**
 * Parses the module map from the app.json packaged into APK assets (key
 * `otaModules`, falling back to `legoModules` for lego-style apps) and
 * orchestrates the background OTA check for every module with `otaUpdates: true`.
 */
object OtaModuleManager {

    @Volatile
    internal var modules: Map<String, OtaModule> = emptyMap()
        private set

    @Volatile
    private var initialized = false

    // Thread-safe map of the bundle version actually resolved per module.
    private val bundleVersionMap = ConcurrentHashMap<String, String>()

    /**
     * Synchronous, local-only app.json parse — safe to call on the launch path.
     * Never throws; on any failure the module map stays empty and OTA is disabled.
     */
    fun init(context: Context, assetFileName: String = "app.json") {
        if (initialized) return
        try {
            val jsonString =
                context.assets.open(assetFileName).bufferedReader().use { it.readText() }
            val root = JSONObject(jsonString)
            val modulesJson = root.optJSONObject("otaModules") ?: root.optJSONObject("legoModules")
            if (modulesJson == null) {
                Log.w("CashifyOTA", "OtaModuleManager::init::no otaModules/legoModules in $assetFileName, OTA disabled")
                initialized = true
                return
            }
            val parsed = mutableMapOf<String, OtaModule>()
            modulesJson.keys().forEach { key ->
                val moduleJson = modulesJson.getJSONObject(key)
                parsed[key] = OtaModule(
                    moduleName = moduleJson.getString("moduleName"),
                    modulePath = moduleJson.getString("modulePath"),
                    configKey = moduleJson.getString("configKey"),
                    scheme = moduleJson.optString("scheme", ""),
                    moduleVersion = moduleJson.getString("moduleVersion"),
                    bundlePriority = moduleJson.optInt("bundlePriority", Int.MAX_VALUE),
                    otaUpdates = moduleJson.optBoolean("otaUpdates", false),
                    launcher = moduleJson.optBoolean("launcher", false),
                )
            }
            modules = parsed.toMap()
            initialized = true
            Log.d("CashifyOTA", "OtaModuleManager::init::modules: ${modules.keys.joinToString(",")}")
        } catch (t: Throwable) {
            Log.e("CashifyOTA", "OtaModuleManager::init failed, OTA disabled: ${t.message}", t)
            modules = emptyMap()
            initialized = true
        }
    }

    /**
     * Fire-and-forget background check: temp cleanup -> Remote Config -> global
     * then per-module safe-mode kill switches -> per-module download-if-needed +
     * stale/rollback cleanup. Downloaded bundles are picked up on the NEXT launch.
     */
    fun loadBundlesAsync(context: Context) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                init(context)
                val otaModules = modules.values.filter { it.otaUpdates }
                if (otaModules.isEmpty()) {
                    Log.d("CashifyOTA", "OtaModuleManager::no OTA modules configured, skipping")
                    return@launch
                }

                OtaBundleManager.cleanupTemporaryBundles(context)

                if (!NetworkUtil.isNetworkAvailable(context)) {
                    Log.d("CashifyOTA", "OtaModuleManager::network not available, skipping bundle check")
                    return@launch
                }

                // Global kill switch: wipes EVERY module and stops the whole check.
                if (OtaRemoteConfig.getEnableSafeMode(context)) {
                    Log.d("CashifyOTA", "OtaModuleManager::global safe mode enabled, clearing all module bundles")
                    otaModules.forEach { OtaBundleManager.cleanupModuleBundles(context, it) }
                    // Persisted locally so the NEXT launch stays on the asset bundle
                    // even when offline.
                    OtaPreferences.setSafeModeEnabled(context, true)
                    return@launch
                }
                OtaPreferences.setSafeModeEnabled(context, false)

                val sortedModules =
                    otaModules.sortedWith(compareBy({ it.bundlePriority }, { it.moduleName }))
                sortedModules.forEach { module ->
                    // Per-module kill switch: wipes ONLY this module; others continue.
                    if (OtaRemoteConfig.getModuleEnableSafeMode(context, module)) {
                        Log.d(
                            "CashifyOTA",
                            "OtaModuleManager::module safe mode enabled for ${module.moduleName}, clearing its bundles"
                        )
                        OtaBundleManager.cleanupModuleBundles(context, module)
                        OtaPreferences.setModuleSafeModeEnabled(context, module.configKey, true)
                        return@forEach
                    }
                    OtaPreferences.setModuleSafeModeEnabled(context, module.configKey, false)

                    Log.d(
                        "CashifyOTA",
                        "OtaModuleManager::checking module: ${module.moduleName}, priority: ${module.bundlePriority}"
                    )
                    OtaBundleManager.downloadBundleIfNeeded(context, module)
                    OtaBundleManager.cleanupStaleBundles(context, module)
                }
            } catch (t: Throwable) {
                Log.e("CashifyOTA", "OtaModuleManager::loadBundlesAsync failed: ${t.message}", t)
            }
        }
    }

    fun launcherModule(): OtaModule? = modules.values.firstOrNull { it.launcher }

    fun getModuleByName(moduleName: String): OtaModule? = modules[moduleName]

    fun getBundleVersion(moduleName: String): String? = bundleVersionMap[moduleName]

    fun setBundleVersion(moduleName: String, version: String) {
        bundleVersionMap[moduleName] = version
    }
}

File: ios/OtaModuleManager.swift
import Foundation

/// Parses the module map from the app.json bundle resource (key `otaModules`,
/// falling back to `legoModules` for lego-style apps) and orchestrates the
/// background OTA check for every module with `otaUpdates: true`.
public final class OtaModuleManager {

  public static let shared = OtaModuleManager()

  private(set) var modules: [String: OtaModule] = [:]
  private var initialized = false

  private let lock = NSLock()
  private var bundleVersionMap: [String: String] = [:]

  private init() {}

  /// Synchronous, local-only app.json parse — safe on the launch path. Never
  /// throws; on any failure the module map stays empty and OTA is disabled.
  public func initModules() {
    guard !initialized else { return }
    initialized = true
    guard let url = Bundle.main.url(forResource: "app", withExtension: "json") else {
      Log.e("OtaModuleManager::init::app.json not found in bundle, OTA disabled")
      return
    }
    do {
      let data = try Data(contentsOf: url)
      guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        Log.e("OtaModuleManager::init::app.json is not an object, OTA disabled")
        return
      }
      let modulesJson = (json["otaModules"] as? [String: [String: Any]])
        ?? (json["legoModules"] as? [String: [String: Any]])
      guard let modulesJson else {
        Log.e("OtaModuleManager::init::no otaModules/legoModules in app.json, OTA disabled")
        return
      }
      var parsed: [String: OtaModule] = [:]
      for (key, moduleJson) in modulesJson {
        guard
          let moduleName = moduleJson["moduleName"] as? String,
          let modulePath = moduleJson["modulePath"] as? String,
          let configKey = moduleJson["configKey"] as? String,
          let moduleVersion = moduleJson["moduleVersion"] as? String
        else {
          Log.e("OtaModuleManager::init::invalid module entry: \(key)")
          continue
        }
        parsed[key] = OtaModule(
          moduleName: moduleName,
          modulePath: modulePath,
          configKey: configKey,
          scheme: moduleJson["scheme"] as? String ?? "",
          moduleVersion: moduleVersion,
          bundlePriority: moduleJson["bundlePriority"] as? Int ?? Int.max,
          otaUpdates: moduleJson["otaUpdates"] as? Bool ?? false,
          launcher: moduleJson["launcher"] as? Bool ?? false
        )
      }
      modules = parsed
      Log.d("OtaModuleManager::init::modules: \(modules.keys.joined(separator: ","))")
    } catch {
      Log.e("OtaModuleManager::init failed, OTA disabled: \(error.localizedDescription)")
      modules = [:]
    }
  }

  /// Fire-and-forget background check: temp cleanup -> Remote Config -> global
  /// then per-module safe-mode kill switches -> per-module download-if-needed +
  /// stale/rollback cleanup. Downloaded bundles are picked up on the NEXT launch.
  public func loadBundlesAsync() {
    Task.detached(priority: .background) {
      await self.loadBundles()
    }
  }

  private func loadBundles() async {
    initModules()
    let otaModules = modules.values.filter { $0.otaUpdates }
    guard !otaModules.isEmpty else {
      Log.d("OtaModuleManager::no OTA modules configured, skipping")
      return
    }

    OtaBundleManager.cleanupTemporaryBundles()

    // Global kill switch: wipes EVERY module and stops the whole check.
    if await OtaRemoteConfig.getEnableSafeMode() {
      Log.d("OtaModuleManager::global safe mode enabled, clearing all module bundles")
      for module in otaModules {
        OtaBundleManager.cleanupModuleBundles(module: module)
      }
      // Persisted locally so the NEXT launch stays on the asset bundle even offline.
      OtaPreferences.setSafeModeEnabled(true)
      return
    }
    OtaPreferences.setSafeModeEnabled(false)

    let sortedModules = otaModules.sorted {
      ($0.bundlePriority, $0.moduleName) < ($1.bundlePriority, $1.moduleName)
    }
    for module in sortedModules {
      // Per-module kill switch: wipes ONLY this module; others continue.
      if await OtaRemoteConfig.getModuleEnableSafeMode(module: module) {
        Log.d("OtaModuleManager::module safe mode enabled for \(module.moduleName), clearing its bundles")
        OtaBundleManager.cleanupModuleBundles(module: module)
        OtaPreferences.setModuleSafeModeEnabled(module.configKey, true)
        continue
      }
      OtaPreferences.setModuleSafeModeEnabled(module.configKey, false)

      Log.d("OtaModuleManager::checking module: \(module.moduleName), priority: \(module.bundlePriority)")
      await OtaBundleManager.downloadBundleIfNeeded(module: module)
      await OtaBundleManager.cleanupStaleBundles(module: module)
    }
  }

  func launcherModule() -> OtaModule? {
    modules.values.first { $0.launcher }
  }

  func getModuleByName(_ moduleName: String) -> OtaModule? {
    modules[moduleName]
  }

  func getBundleVersion(_ moduleName: String) -> String? {
    lock.lock()
    defer { lock.unlock() }
    return bundleVersionMap[moduleName]
  }

  func setBundleVersion(moduleName: String, version: String) {
    lock.lock()
    defer { lock.unlock() }
    bundleVersionMap[moduleName] = version
  }
}
